go/vrf/ecvrf.go

```diff
@@ -43,3 +44,29 @@ func NewKey(curve elliptic.Curve, sk []byte) *PrivateKey {
 		d:         new(big.Int).SetBytes(sk),           // Use SK to derive the VRF secret scalar x
 	}
 }
+
+// ECVRFParams holds shared values across ECVRF implementations.
+// ECVRFParams also has generic algorithms that rely on ECVRFAux for specific sub algorithms.
+type ECVRFParams struct {
+	suiteString []byte         // Single nonzero octet specifying the ECVRF ciphersuite
+	ec          elliptic.Curve // Elliptic curve defined over F
+	n           int            // 2n is the length, in bytes, of a field element in F
+	ptLen       int            // Length, in octets, of an EC point encoded as an octet string
+	qLen        int            // Length of the prime order of the EC group in octets. (Typically ~2n)
+	cofactor    byte           // The number of points on EC divided by the prime order of the group
+	hash        crypto.Hash    // Cryptographic hash function
+	aux         ECVRFAux       // Suite specific helper functions
+}
+
+// ECVRFAux contains auxiliary functions nessesary for the computation of ECVRF.
+type ECVRFAux interface {
+	// PointToString converts an EC point to an octet string.
+	PointToString(Px, Py *big.Int) []byte
+
+	// StringToPoint converts an octet string to an EC point.
+	// This function MUST output INVALID if the octet string does not decode to an EC point.
+	StringToPoint(h []byte) (Px, Py *big.Int, err error)
+
+	// ArbitraryStringToPoint converts an arbitrary 32 byte string s to an EC point.
+	ArbitraryStringToPoint(s []byte) (Px, Py *big.Int, err error)
```

Comment:
From the descriptions it is not clear to me what the difference to `StringToPoint` is here (what is the limitation for `StringToPoint`?). Does this also output invalid if the string does not decode to an EC point?

Reply:
The basic difference is that `StringToPoint` == `compressedUnmarshal`, while `ArbitraryStringToPoint` prepends `0x02` to interpret an arbitrary string as a compressed curve point.